File: tools/interp/timeline.py
"""Turn a resolved capture timeline into an exact 60 fps sample plan.

Every sample time is n * 1000 / fps milliseconds, held as a Fraction so no
rounding creeps in. A sample that lands exactly on an authored pose time copies
that pose. A sample after the last pose copies the last pose, which is the hold.
A sample inside a transition asks RIFE for s = (t - t0) / (t1 - t0). A cut named
in `cuts` stops interpolation into that pose, so the earlier pose stays on
screen until the cut time.
"""

from fractions import Fraction

COPY = "copy"
HOLD = "hold"
INTERP = "interp"


def sample_count(end_ms, fps):
    """Number of samples: ceil(end_ms * fps / 1000)."""
    total = Fraction(end_ms) * fps / 1000
    whole = total.numerator // total.denominator
    return whole + (1 if total.denominator != 1 or total != whole else 0)


def plan(spec, fps):
    """Return one entry per output frame, in time order.

    Each entry is a dict with `t` (Fraction milliseconds), `kind`, `source`
    (one pose index for a copy or a hold, two for an interpolation) and `s`
    (a Fraction for an interpolation, otherwise None).
    """
    poses = spec["poses"]
    times = [Fraction(p["t_ms"]) for p in poses]
    cuts = set(spec.get("cuts") or [])
    end = Fraction(spec["end_ms"])
    step = Fraction(1000, fps)

    entries = []
    k = 0
    for n in range(sample_count(spec["end_ms"], fps)):
        t = n * step
        while k + 1 < len(poses) and times[k + 1] <= t:
            k += 1
        if t == times[k]:
            entries.append({"t": t, "kind": COPY, "source": [poses[k]["index"]], "s": None})
        elif k + 1 >= len(poses):
            entries.append({"t": t, "kind": HOLD, "source": [poses[k]["index"]], "s": None})
        elif poses[k + 1]["index"] in cuts:
            entries.append({"t": t, "kind": HOLD, "source": [poses[k]["index"]], "s": None})
        else:
            s = (t - times[k]) / (times[k + 1] - times[k])
            entries.append({
                "t": t,
                "kind": INTERP,
                "source": [poses[k]["index"], poses[k + 1]["index"]],
                "s": s,
                "pose": k,
            })
    for i, entry in enumerate(entries):
        nxt = entries[i + 1]["t"] if i + 1 < len(entries) else end
        entry["duration"] = nxt - entry["t"]
    return entries


def uniform_gap(spec):
    """Return the pose spacing in milliseconds when every gap is the same.

    Returns None when the poses are not evenly spaced, which sends the pipeline
    down the one-process-per-sample path.
    """
    times = [p["t_ms"] for p in spec["poses"]]
    if len(times) < 2:
        return None
    gap = times[1] - times[0]
    if gap <= 0:
        return None
    for a, b in zip(times, times[1:]):
        if b - a != gap:
            return None
    return Fraction(gap)

File: src/interp/timeline.cpp
#include "interp/timeline.h"

#include <set>

namespace swchess::interp {

const char* frameKindName(FrameKind kind) {
    switch (kind) {
        case FrameKind::Blank: return "blank";
        case FrameKind::Copy: return "copy";
        case FrameKind::Hold: return "hold";
        case FrameKind::Interp: return "interp";
    }
    return "blank";
}

std::int64_t sampleCount(std::int64_t endMs, int fps) {
    return (Fraction(endMs) * Fraction(fps, 1000)).ceiling();
}

std::vector<PlannedFrame> planFrames(const Json& spec, int fps) {
    const Json& poses = spec.at("poses");
    std::vector<Fraction> times;
    times.reserve(poses.size());
    for (const Json& pose : poses) {
        times.emplace_back(pose.at("t_ms").get<std::int64_t>());
    }
    std::set<int> cuts;
    if (spec.contains("cuts") && !spec.at("cuts").is_null()) {
        for (const Json& cut : spec.at("cuts")) {
            cuts.insert(cut.get<int>());
        }
    }
    const std::int64_t endMs = spec.at("end_ms").get<std::int64_t>();
    const Fraction step(1000, fps);

    std::vector<PlannedFrame> frames;
    std::size_t k = 0;
    const std::int64_t total = sampleCount(endMs, fps);
    for (std::int64_t n = 0; n < total; ++n) {
        const Fraction t = Fraction(n) * step;
        PlannedFrame frame;
        frame.t = t;
        if (times.empty() || t < times[0]) {
            frame.kind = FrameKind::Blank;
            frames.push_back(frame);
            continue;
        }
        while (k + 1 < poses.size() && times[k + 1] <= t) {
            ++k;
        }
        const int currentIndex = poses[k].at("index").get<int>();
        if (t == times[k]) {
            frame.kind = FrameKind::Copy;
            frame.source = {currentIndex};
        } else if (k + 1 >= poses.size()) {
            frame.kind = FrameKind::Hold;
            frame.source = {currentIndex};
        } else if (cuts.count(poses[k + 1].at("index").get<int>()) != 0) {
            frame.kind = FrameKind::Hold;
            frame.source = {currentIndex};
        } else {
            frame.kind = FrameKind::Interp;
            frame.source = {currentIndex, poses[k + 1].at("index").get<int>()};
            frame.step = (t - times[k]) / (times[k + 1] - times[k]);
            frame.hasStep = true;
            frame.pose = k;
        }
        frames.push_back(frame);
    }

    const Fraction end(endMs);
    for (std::size_t i = 0; i < frames.size(); ++i) {
        const Fraction next = i + 1 < frames.size() ? frames[i + 1].t : end;
        frames[i].duration = next - frames[i].t;
    }
    return frames;
}

Json soundEvents(const Json& spec) {
    Json out = Json::array();
    for (const Json& pose : spec.at("poses")) {
        if (!pose.contains("sound") || pose.at("sound").is_null()) {
            continue;
        }
        const Json& sound = pose.at("sound");
        Json entry = Json::object();
        entry["pose"] = pose.at("index");
        // Older files wrote no t_ms inside the sound, and for those the pose
        // time is the best guess left.
        if (sound.contains("t_ms") && !sound.at("t_ms").is_null()) {
            entry["t_ms"] = sound.at("t_ms");
        } else {
            entry["t_ms"] = pose.at("t_ms");
        }
        entry["sound"] = sound;
        out.push_back(entry);
    }
    return out;
}

}  // namespace swchess::interp
